rust/worker/src/segment/metadata_segment.rs

```diff
@@ -0,0 +1,1158 @@
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        let mut u32_metadata_index_writer = self
+            .u32_metadata_index_writer
+            .take()
+            .ok_or_else(|| MetadataSegmentError::NoWriter)?;
+        let res = u32_metadata_index_writer.write_to_blockfile().await;
+        self.u32_metadata_index_writer = Some(u32_metadata_index_writer);
+        match res {
+            Ok(_) => {}
+            Err(_) => return Err(MetadataSegmentError::BlockfileWriteError),
+        }
+
+        Ok(())
+    }
+}
+
+impl SegmentWriter for MetadataSegmentWriter {
+    fn apply_materialized_log_chunk(
+        &self,
+        records: crate::execution::data::data_chunk::Chunk<MaterializedLogRecord>,
+    ) {
+        for record in records.iter() {
+            let segment_offset_id = record.0.segment_offset_id;
+            match record.0.log_record.record.operation {
+                Operation::Add => {
+                    match &record.0.materialized_record.metadata {
+                        Some(metadata) => {
+                            for (key, value) in metadata.iter() {
+                                match value {
+                                    MetadataValue::Str(value) => {
+                                        match &self.string_metadata_index_writer {
+                                            Some(writer) => {
+                                                let _ = writer.set(
+                                                    key,
+                                                    value.as_str(),
+                                                    segment_offset_id,
+                                                );
+                                            }
+                                            None => {}
+                                        }
+                                    }
+                                    MetadataValue::Float(value) => {
+                                        match &self.f32_metadata_index_writer {
+                                            Some(writer) => {
+                                                let _ = writer.set(
+                                                    key,
+                                                    *value as f32,
+                                                    segment_offset_id,
+                                                );
+                                            }
+                                            None => {}
+                                        }
+                                    }
+                                    MetadataValue::Int(value) => {
+                                        match &self.u32_metadata_index_writer {
+                                            Some(writer) => {
+                                                let _ = writer.set(
```

Comment:
Not related to this diff but one thing that came to my mind is that we don't write atomically to the segments today i.e. it can happen that the compactor node crashes after writing for e.g. to the record segment but before writing to the metadata segment, etc.

Reply:
Capturing our conversation here for posterity:

Yes, that can happen. We designed for this and that's why the flush protocol is: write new blockfiles, update sysdb to point to new blockfiles, purge logs. If we crash anywhere in there we're still correct.